### ISSUE TITLE
Ship OpenAPI Specification not working

### ISSUE THREAD

**Issue by kdurkin77** [UPS-API/api-documentation#3]
Is there a newer Ship API OpenAPI specification? It doesn't appear to work as is. The response json is not deserializable - there are a few fields that are arrays in the specification that come over as objects and visa versa. For instance, PackageResults in the spec is an array but it comes over as an object. I am using the dev environment to test this, not sure if it's different in prod

**Comment by kdurkin77** [comment]
Additionally, PackageResults.ItemizedCharges sometimes comes over as an object and other times as an array.

**Comment by kdurkin77** [comment]
It also looks like PackageResults.ShippingLable.GraphicImagePart is required but is not being returned

**Comment by chuntoulee** [comment]
Thank you for bringing this concern to our attention.  Our technical team is looking into it and will provide a more substantive response along with resolution shortly.

**Comment by caleblanchard** [comment]
Is there any update on this? We are trying to implement the REST API and we are seeing the same issues. There are parts that sometimes come back as an object and sometimes as an array. It is not consistent.

**Comment by chuntoulee** [comment]
Thank you for the follow-up.  We plan to publish the updated Open API spec by 07/07/2023, if not sooner...

**Comment by caleblanchard** [comment]
> Thank you for the follow-up. We plan to publish the updated Open API spec by 07/07/2023, if not sooner...

I have not seen an update to the Open API spec. Can you provide an update as to when it will be updated?

**Comment by sjlehn** [comment]
They pushed an update to the spec yesterday, but there's still validation issues in the Rate API, and some validation and schema issues in Ship...I was able to clean it up with a formatter and a bit of `sed`, but they really should work out of the box

**Comment by kdurkin77** [comment]
Yeah we were able to get it working in .NET by using a json converter and adding it using UpdateJsonSerializerSettings for 3 different types but it really should just work out of the box like you said. I don't know that open API supports types that change between an object and an array. Most APIs just use an array with a single item it in, not sure why they went a different way here

**Comment by caleblanchard** [comment]
I agree with @kdurkin77. You can't have a type that switches between an object and an array. It needs to be an array. Sometimes that array will only have one element, but it is not supported to sometimes return an array and sometimes return an object. I am currently seeing this issue in the rating service with the RatedShipment Array. If there is only one RatedShipment, then it is returning as an object and not as an array with one element.

**Comment by david-szabo97** [comment]
UPS removed the possibility of obtaining "legacy" access keys and released this API specification that's full of errors. That is an excellent move from such a big company...

@chuntoulee 
Do UPS have any plans on fixing the array/object issue, or do we need to rely on custom serialization logic to force them to array?

**Comment by elidrissidev** [comment]
+1, this issue is also present in the Address Validation API.

**Comment by ericvandamme** [comment]
In C#, I made some code to fix the promise of an array but getting a single object. It works with C# and Newtonsoft. It fixes the problem so if you're using NSwag to generate your code from the swagger document, this will speed things along the development process.

```
    public class BadUPSCollectionConverter<T> : JsonConverter {

        public override bool CanConvert(Type objectType)
            => objectType.IsAssignableTo(typeof(ICollection<T>));

        public override object? ReadJson(JsonReader reader, Type objectType, object? existingValue, JsonSerializer serializer) {
            JToken token = JToken.Load(reader);
            if (token.Type == JTokenType.Array) {
                var list = new List<T>();
                foreach (var t in token.Children()) {
                    T? instance = t.ToObject<T>(serializer);
                    if (instance != null)
                        list.Add(instance);
                }
                return list;
            } else if (token.Type == JTokenType.Object) {
                T? instance = token.ToObject<T>(serializer);
                return instance != null
                    ? new List<T>() { instance }
                    : [];
            } else throw new JsonSerializationException("Unexpected token type: " + token.Type);
        }

        public override bool CanWrite => false;

        public override void WriteJson(JsonWriter writer, object? value, JsonSerializer serializer) {
            throw new NotSupportedException();
        }
    }
```

Then I simply add the converters for all the types that promise to be arrays.

```
        static partial void UpdateJsonSerializerSettings(Newtonsoft.Json.JsonSerializerSettings settings) {
            settings.Converters.Add(new BadUPSCollectionConverter<Shipment_ShipmentIndicationType>());
            settings.Converters.Add(new BadUPSCollectionConverter<Shipment_Package>());
            settings.Converters.Add(new BadUPSCollectionConverter<PaymentDetails_ShipmentCharge>());
            settings.Converters.Add(new BadUPSCollectionConverter<FreightDensityInfo_HandlingUnits>());
            settings.Converters.Add(new BadUPSCollectionConverter<HazMat_HazMatChemicalRecord>());
            settings.Converters.Add(new BadUPSCollectionConverter<DeliveryTimeInformation_ReturnContractServices>());
            settings.Converters.Add(new BadUPSCollectionConverter<RateResponse_RatedShipment>());
            settings.Converters.Add(new BadUPSCollectionConverter<Response_Alert>());
            settings.Converters.Add(new BadUPSCollectionConverter<Response_AlertDetail>());
            settings.Converters.Add(new BadUPSCollectionConverter<ElementLevelInformation_ElementIdentifier>());
            settings.Converters.Add(new BadUPSCollectionConverter<RatedShipment_Disclaimer>());
            settings.Converters.Add(new BadUPSCollectionConverter<RatedShipment_RatedShipmentAlert>());
            settings.Converters.Add(new BadUPSCollectionConverter<RatedShipment_ItemizedCharges>());
            settings.Converters.Add(new BadUPSCollectionConverter<RatedShipment_TaxCharges>());
            settings.Converters.Add(new BadUPSCollectionConverter<RatedShipment_RatedPackage>());
            settings.Converters.Add(new BadUPSCollectionConverter<FRSShipmentData_HandlingUnits>());
            settings.Converters.Add(new BadUPSCollectionConverter<NegotiatedRateCharges_ItemizedCharges>());
            settings.Converters.Add(new BadUPSCollectionConverter<NegotiatedRateCharges_TaxCharges>());
            settings.Converters.Add(new BadUPSCollectionConverter<RatedPackage_Accessorial>());
            settings.Converters.Add(new BadUPSCollectionConverter<RatedPackage_ItemizedCharges>());
            settings.Converters.Add(new BadUPSCollectionConverter<RatedPackage_RateModifier>());
            settings.Converters.Add(new BadUPSCollectionConverter<NegotiatedCharges_ItemizedCharges>());
            settings.Converters.Add(new BadUPSCollectionConverter<TimeInTransit_ServiceSummary>());
        }
```

**Comment by UPSRahulBhatnagar** [comment]
Hello all, we understand the issue being presented here. The root cause is due to legacy code behind the rest service. We have been evaluating ways to fix this, but due to a large number of customers using these APIs we do not want to break backwards compatibility for those who have coded around the issue. We will be coming up with a new way to handle arrays appropriately moving forward and will publish this on the developer portal once ready.

In the meantime, a solution some of our customers have implemented is a response filter that will investigate elements prior to serialization. Elements in a response object that can be an array are then tagged with the filter, and wrapped in a list if only a single instance is returned.

**Comment by DemigodCode** [comment]
Hi Rahul,

legacy code, yes. You use a version in the header information of a request. Why don't you just introduce a new version "2" instead of "1.x". I really don't understand why customers should care about a fix.

Using swagger we can create all elements returned and send to the API automatically. And we just have to provide a HttpClient. Now I have two options:
- Extend the HttpClient and overwrite the response handling. For that I have to copy code to change 3 lines in it and have to keep it up to date by myself.

So, again.

There are two solutions which will make it possible for us all to use the API again:
- Introduce a new version of the API
- Modify the Ship.json that it can return "anyOf" PackageResults[] or just the properties of a PackageResult

Both won't break any backwards compatibility.

UPS forces us to move from a working SOAP to a broken REST-API that doesn't work as expected or documented until July '24. Until then I expect a version that will work.

**Comment by UPSRahulBhatnagar** [comment]
Hi, thank you for your comment. The changes in the OpenAPI documentation were to illustrate how the APIs currently function. Moving forward, we will determine a new way to properly handle arrays and, once ready, will publish this on the developer portal.

**Comment by DemigodCode** [comment]
Any updates on this? Or something like a maillist where we can sign on to get actively informed once it's fixed?

Waiting for the fix to go on integrating.

**Comment by florentpln** [comment]
Is there update on this? For sure using oneOf is not a very good idea until it is on openapi generator roadmap : https://openapi-generator.tech/docs/roadmap/

**Comment by DemigodCode** [comment]
I'm not very confident that UPS is trying to fix this. And we're getting closer to the SOAP EOL

**Comment by AlexAtPeerless** [comment]
At this point I suggest everybody aware of this issue file a complaint to their UPS representative about UPS's lack of readiness for their own June 2024 deadline for the REST API adoption.
UPS cannot expect in all seriousness all of its customers to use an API/documentation that's not reasonably usable.
The handling of arrays is nonsensical and we shouldn't have to do R&D to simply consume web services.

**Comment by ipavkovic** [comment]
java people could change their objectmapper like this
`
return JsonMapper.builder()
        .configure(DeserializationFeature.ACCEPT_SINGLE_VALUE_AS_ARRAY, true)
        .build();
`, manipulate openapi spec from ups to only have lists instead of non-list alternative and give it a go with openapi-generator.

**Comment by NXTwoThou** [comment]
I got directed here because I have an old ASP.NET v4.8 Web Site project that needed to move from SOAP to the REST API.   We're using System.Text.Json(v8.0.2).  At some point, the VB.NET folks lost the ability to use JsonConverter's due to Utf8JsonReader.  The response to this was to tell VB.NET people to write C# libraries to handle JsonConverters.  Unfortunately, for whatever reason, I'm getting CS0619 errors for Utf8JsonReader even though everything in the project is C#.  After a lot of frustration and searching I found an alternate way of taking care of the issue and adapted it to my code for UPS.  I figured I'd share it in case someone else gets directed here and is unable to use JsonConverters for whatever reason(VB.NET people can adapt this technique too).

I've created three converters:

`
using System.Text.Json;

public class RESTClient {
public static List<T> ObjectToListJsonConverter<T>(object o)
{
    if (o == null)
        return null;
    if(o is JsonElement)
    {
        JsonElement element = (JsonElement)o;
        switch (element.ValueKind)
        {
            case JsonValueKind.Array:
                return JsonSerializer.Deserialize<List<T>>(element.GetRawText());
            case JsonValueKind.String:
                {
                    string value=element.GetString();
                    if (value == null || value.Length == 0)
                        return null;
                    return new List<T>() { (T)Convert.ChangeType(value, typeof(T)) };
                }
            case JsonValueKind.Object:
                return new List<T>() { JsonSerializer.Deserialize<T>(element.GetRawText()) };
        }
    }
    return (List<T>)o;
}

public static T ObjectToClassJsonConverter<T>(object o) where T : class
{
    if (o == null)
        return null;
    if(o is JsonElement)
    {
        JsonElement element = (JsonElement)o;
        switch (element.ValueKind)
        {
            case JsonValueKind.String:
                {
                    string value=element.GetString();
                    if (value == null || value.Length == 0 || value.ToUpperInvariant()=="NULL")
                        return null;
                    return (T)Convert.ChangeType(value, typeof(T));
                }
            case JsonValueKind.Object:
                return JsonSerializer.Deserialize<T>(element.GetRawText());
        }
    }
    return (T)o;
}

// I'm not using, but someone else might need something so might as well write it.
public static T[] ObjectToArrayJsonConverter<T>(object o)
{
    if (o == null)
        return null;
    if(o is JsonElement)
    {
        JsonElement element = (JsonElement)o;
        switch (element.ValueKind)
        {
            case JsonValueKind.Array:
                return JsonSerializer.Deserialize<T[]>(element.GetRawText());
            case JsonValueKind.String:
                {
                    string value=element.GetString();
                    if (value == null || value.Length == 0)
                        return null;
                    var t = new T[1];
                    t[0] = (T)Convert.ChangeType(value, typeof(T));
                    return t;
                }
            case JsonValueKind.Object:
                {
                    var t = new T[1];
                    t[0] = JsonSerializer.Deserialize<T>(element.GetRawText());
                    return t;
                }                    
        }
    }
    return (T[])o;
}

}
`

You then take whatever property you had that's being poorly emitted by UPS.  Decorate it with JsonIgnore.  Create a generic object for it and use JsonPropertyName to cast this object to the name you need to (de)serialize as.  Then change the get/set on the JsonIgnored one to run one of the converters.

Here's three examples of how they are utilized:

`
// Instead of public List<string> AddressLine { get; set; }
[JsonPropertyName("AddressLine")]
public object AddressLineObject { get; set; }
[JsonIgnore]
public List<string> AddressLine { get { return RESTClient.ObjectToListJsonConverter<string>(AddressLineObject); } set { AddressLineObject = value; } }

// Instead of public List<UPSRestAlert> Alert { get; set; }
[JsonPropertyName("Alert")]
public object AlertObject { get; set; }
[JsonIgnore]
public List<UPSRestAlert> Alert { get { return RESTClient.ObjectToListJsonConverter<UPSRestAlert>(AlertObject); } set { AlertObject = value; } }

// Instead of public UPSRestReceipt Receipt { get; set; }
[JsonPropertyName("Receipt")]
public object ReceiptObject { get; set; }
[JsonIgnore]
public UPSRestReceipt Receipt
{
    get { return RESTClient.ObjectToClassJsonConverter<UPSRestReceipt>(ReceiptObject); }
    set { ReceiptObject = value; }
}
`

As I was using https://json2csharp.com/ to convert the structures from the documentation.  Anytime I had a List I would rewrite things.  While testing, discovered the -other- nasty bit where UPS would return "Receipt":"" rather than "Receipt":{} or "Receipt":null or not include at all.  So that's why I had to create the ObjectToClassJsonConverter to handle having a 0 length string being presented in the json.

I only started on Tuesday, but I've gotten all we use completed(/labels/recovery,  /addressvalidation, /shipments/ship(to and return label), /pickupcreation/pickup) using this technique.

**Comment by DemigodCode** [comment]
Hey UPS,
not every language has this support.

Got an email (2 months left, get ready)... So... when will you get ready for your own deadline?

**Comment by DemigodCode** [comment]
> Would also love to know if there's going to be any movement on this - many languages are not set up to handle this arbitrary return of objects or arrays of objects. This is a fairly critical impediment to adoption, and there's very little time remaining before the cutoff.

@UPSRahul we are going to need any information UPS will handle this now. The deadline is in only 2 months. We have to implement the new API and test it. UPS needs to communicate the next steps!

**Comment by ericrosfelder** [comment]
> > Would also love to know if there's going to be any movement on this - many languages are not set up to handle this arbitrary return of objects or arrays of objects. This is a fairly critical impediment to adoption, and there's very little time remaining before the cutoff.
> 
> @UPSRahul we are going to need any information UPS will handle this now. The deadline is in only 2 months. We have to implement the new API and test it. UPS needs to communicate the next steps!

Yes, @UPSRahul we need an answer on this too.

I have given simple steps to reproduce the errors in this issue:
https://github.com/UPS-API/java-api-examples/issues/8

I am in direct contact with UPS and it seems that no one really wants to address the root issue.

**Comment by UPSRahul** [comment]
Hi, thank you for your patience. The APIs that exhibit this behavior have been updated with new versions. For the previously impacted fields, the new API versions will always return an array of multiple objects or an array of one object. The new Shipping, Rating, Pickup, and DangerousGoods (Chemical Reference Data) APIs will use v2403 while the DangerousGoods (Acceptance Audit Pre-check), Pre Notification, Paperless Documents, Locator, Quantum View, Ship (Void), Address Validation APIs will use v2. Any APIs not mentioned were not impacted by this issue.

The GitHub will be updated on 4/3/24 with the new documentation, and the developer portal will be updated shortly. 

Thank you again for your patience, and please let us know if there are any questions or concerns with the new API versions.

**Comment by ipavkovic** [comment]
@UPSRahul that sound good. When will you upload the newest openapi specs in this repository?

**Comment by ataylorkato** [comment]
@UPSRahul As a suggestion, would you consider utilizing the OpenAPI info.version field in the specs going forward? It would be helpful.

**Comment by UPSRahul** [comment]
Hi all, the updated specs have been pushed.

**Comment by AlexAtPeerless** [comment]
Hello, thanks for the new versions.
I noticed that there is still oneOf discriminators in the Shipping new yaml for two elements:
```
LabelResults
TrackingCandidate
```
As the use of oneOf in the previous version indicated the presence of the issue with the response containing either an Object or an Array for those elements, I suggest they should be changed to always be arrays like the other corrections.
Also the use of oneOf causes problems for many clients/technologies so it's one reason more to remove them. (This should also obvisouly apply to other APIs.)

**Comment by NXTwoThou** [comment]
> Hi all, the updated specs have been pushed.

Does there happen to be a single list of latest /{version}/ and SubVersion for each endpoint anywhere?

**Comment by UPSRahul** [comment]
> Hello @UPSRahul , thanks for the new versions. I noticed that there is still oneOf discriminators in the Shipping new yaml for two elements:
> 
> ```
> LabelResults
> TrackingCandidate
> ```
> 
> As the use of oneOf in the previous version indicated the presence of the issue with the response containing either an Object or an Array for those elements, I suggest they should be changed to always be arrays like the other corrections. Also the use of oneOf causes problems for many clients/technologies so it's one reason more to remove them. (This should also obvisouly apply to other APIs.)

Hi, please reference issue #81 for more detail.

**Comment by UPSRahul** [comment]
> > Hi all, the updated specs have been pushed.
> 
> Does there happen to be a single list of latest /{version}/ and SubVersion anywhere?
> 
> For example:
> 
> EndPoint,Version /shipments/,v2403 /pickupcreation/,v2403
> 
> Object,SubVersion ShipmentRequest_Request,2205 LabelRecoveryRequest_Request,1903 PickupCreationRequest_Request,2007

Hi, below is a table of the updated API versions.
![image](https://github.com/UPS-API/api-documentation/assets/147546331/b12505d6-5ac8-4126-9e25-3ac0a59d9296)

**Comment by NXTwoThou** [comment]
> Hi, below is a table of the updated API versions. 

Thanks!  Is Label Recovery still v1903?  

Is there a similar chart for SubVersions?

**Comment by UPSRahul** [comment]
> > Hi, below is a table of the updated API versions.
> 
> Thanks! Is Label Recovery still v1903?
> 
> Is there a similar chart for SubVersions?

There is no new version for Label Recovery, please reference https://github.com/UPS-API/api-documentation/issues/3#issuecomment-2037926051

**Comment by UPSRahul** [comment]
Hi, we are closing this issue as complete. If there are any new questions, please create a new issue or reach out to the UPS Help Desk https://developer.ups.com/contactus
